libs/util/AdvancedMath.arucas

```diff
@@ -0,0 +1,915 @@
+        cos = Math.cos(angle);
+        sin = Math.sin(angle);
+
+        x = this.x - center.x;
+        y = this.y - center.y;
+
+        x = x * cos - y * sin + center.x;
+        y = x * sin + y * cos + center.y;
+        return new Vec2d(x,y);
+    }
+
+    fun set(v) {
+        this.x = v.x;
+        this.y = v.y;
+        return this;
+    }
+
+    fun setX(x) {
+        this.x = x;
+        return this;
+    }
+
+    fun setY(y) {
+        this.y = y;
+        return this;
+    }
+
+    fun setLength (length) {
+        return this.normalize().multiplyScalar(length);
+    }
+
+    fun sub(v) {
+        x = this.x - v.x;
+        y = this.y - v.y;
+        return new Vec2d(x,y);
+    }
+    operator [] (index){
+        return [this.x,this.y][index];
+    }
+    
+    fun toString(){
+        return "[" + this.x + ", " + this.y + "]";
+    }
+
+    fun floor(){
+        return new Vec2d(Math.floor(this.x),Math.floor(this.y));
+    }
+}
+
+class ComplexNumber{
+
+    var real;
+    var imaginary;
+    var arg;
+    var abs;
+
+    //"Constructors"
+    ComplexNumber(r : Number, i : Number){
+        this.real = r;
+        this.imaginary = i;
+        this.arg = Math.arctan(i/r);
+        this.abs = Math.sqrt(r*r + i*i);
+    }
+
+    ComplexNumber(polar : List){
+        this.abs = polar[0];
+        this.arg = polar[1];
+        this.real = this.abs * Math.cos(this.arg);
+        this.imaginary = this.abs * Math.sin(this.arg);
+    }
+
+    static fun complex(r : Number, i : Number): ComplexNumber{
+        return new ComplexNumber(r,i);
+    }
+
+
+    // Operators
+    operator +(other){
+        if(Type.of(other).getName() == 'Number'){
+            return this.add(other,0);
+        }
+        else if(Type.of(other).getName() == 'ComplexNumber'){
+            return this.add(other);
+        }
+        else{
+            return null;
+        }
+    }
+
+    operator -(other){
+        if(Type.of(other).getName() == 'Number'){
+            return this.substract(other,0);
+        }
+        else if(Type.of(other).getName() == 'ComplexNumber'){
+            return this.substract(other);
+        }
+        else{
+            return null;
+        }
+    }
+
+    operator *(other){
+        if(Type.of(other).getName() == 'Number'){
+            return this.multiply(other,0);
+        }
+        else if(Type.of(other).getName() == 'ComplexNumber'){
+            return this.multiply(other);
+        }
+        else{
+            return null;
+        }
+    }
+
+    operator /(other){
+        if(Type.of(other).getName() == 'Number'){
+            return this.multiply(1/other,0);
+        }
+        else if(Type.of(other).getName() == 'ComplexNumber'){
+            return this.divide(other);
+        }
+        else{
+            return null;
+        }
+    }
+
+    operator ^(other){
+        if(Type.of(other).getName() == 'Number'){
+            return this.exponent(other);
+        }
+        else{
+            return null;
```

Comment:
Should you really be returning null here? I feel like that will just lead to confusion, I would throw an error, or even better type hint the parameter in which case it will guarentee the type you can even allow multiple types like `fun s(other: Number | ComplexNumber)`, this applies to a few of the methods here.

Reply:
I meant to comment on the parameters, but apparently I clicked wrong

Reply:
delegated this to surf by tagging on discord